File: vkatz-lib/src/main/java/by/vkatz/utils/SerializableUtils.java
package by.vkatz.utils;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;

/**
 * Created by vKatz
 */
public class SerializableUtils {
    public static boolean commit(File file, Object data) {
        try {
            ObjectOutputStream out = new ObjectOutputStream(new FileOutputStream(file));
            out.writeObject(data);
            out.flush();
            out.close();
            return true;
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
    }

    @SuppressWarnings("unchecked")
    public static <T> T restore(File file, Functions.Func0<T> creator) {
        try {
            return (T) new ObjectInputStream(new FileInputStream(file)).readObject();
        } catch (Exception e) {
            e.printStackTrace();
            try {
                return creator.execute();
            } catch (Exception e1) {
                e1.printStackTrace();
                return null;
            }
        }
    }

    @SuppressWarnings("unchecked")
    public static <T> T restoreOrThrow(File file) throws Exception {
        return (T) new ObjectInputStream(new FileInputStream(file)).readObject();
    }

    public static <T> long generateSerialVersionUID(Class<T> t) {
        long val = 5611235136846231L;
        String fullName = t.getName();
        for (int i = 0; i < fullName.length(); i++)
            val = 31 * val + fullName.charAt(i);
        return val;
    }
}

File: vkatz-lib/src/main/kotlin/by/vkatz/utils/SerializableUtils.kt
package by.vkatz.utils

import java.io.*

object SerializableUtils {
    fun commit(file: File, data: Any): Boolean {
        return try {
            ObjectOutputStream(FileOutputStream(file)).use { it.writeObject(data) }
            true
        } catch (e: Exception) {
            LogUtils.e("SerializableUtils", "Commit error:", e)
            false
        }
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> restore(file: File, creator: () -> T): T {
        return try {
            ObjectInputStream(FileInputStream(file)).use { it.readObject() as T }
        } catch (e: Exception) {
            LogUtils.e("SerializableUtils", "Restore error:", e)
            creator()
        }
    }

    @Throws(Exception::class)
    @Suppress("UNCHECKED_CAST")
    fun <T> restoreOrThrow(file: File): T = ObjectInputStream(FileInputStream(file)).use { it.readObject() as T }
}
